### ISSUE TITLE
Native image compilation throws OOM for JMH benchmark suite

### ISSUE THREAD

**Issue by gigiblender** [oracle/graal#6309]
I am trying to compile with the native image in GraalVM CE 17 22.3.0 the JMH benchmarks at https://github.com/ionutbalosin/jvm-performance-benchmarks. 

When trying to compile the whole suite, after some time, I run into an OOM exception thrown by a forked process. I run with `-J-Xmx32G`. Trying to set a larger heap (e.g. 60G) will cause an exception in the canonicalizer.

```
[4/7] Parsing methods...      [*]                                                                        (1.0s @ 3.69GB)
Exception in thread "native-image pid watcher" java.lang.OutOfMemoryError: GC overhead limit exceeded
```

The full native-image output is:
```
$ $JAVA_HOME/bin/native-image -J-Xmx32G --add-modules=jdk.incubator.vector --initialize-at-build-time=org.openjdk.jmh.infra,org.openjdk.jmh.util.Ut[0/42477]penjdk.jmh.runner.InfraControl,org.openjdk.jmh.runner.InfraControlL0,org.openjdk.jmh.runner.InfraControlL1,org.openjdk.jmh.runner.InfraControlL2,org.openjdk.jmh.runner.InfraControlL3,org.openjdk.jmh.runner.InfraControlL4 -H:+ReportExceptionStackTraces -jar benchmarks/target/benchmarks.jar -cp . graal17ce-image
WARNING: Using incubator modules: jdk.incubator.vector
========================================================================================================================
GraalVM Native Image: Generating 'graal17ce-image' (executable)...
========================================================================================================================
Warning: Could not resolve sun.misc.Contended for reflection configuration. Reason: java.lang.ClassNotFoundException: sun.misc.Contended.
[1/7] Initializing...                                                                                    (2.3s @ 0.33GB)
 Version info: 'GraalVM 22.3.0 Java 17 CE'
 Java version info: '17.0.5+8-jvmci-22.3-b08'
 C compiler: gcc (linux, x86_64, 9.4.0)
 Garbage collector: Serial GC
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced by the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced b
y the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced b
y the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced b
y the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced b
y the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.runner.InfraControl. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced b
y the call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.infra.Blackhole. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced by th
e call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.infra.Blackhole. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced by th
e call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
Warning: RecomputeFieldValue.FieldOffset automatic substitution failed. The automatic substitution registration was attempted because a call to sun.misc.Unsafe.objectFieldOffset(Field) was detected in the static
 initializer of org.openjdk.jmh.infra.Blackhole. Detailed failure reason(s): The argument of Unsafe.objectFieldOffset(Field) is not a constant field., Could not determine the field where the value produced by th
e call to sun.misc.Unsafe.objectFieldOffset(Field) for the field offset computation is stored. The call is not directly followed by a field store or by a sign extend node followed directly by a field store. 
[2/7] Performing analysis...  [*********]                                                               (11.7s @ 3.10GB)
   7,412 (87.31%) of  8,489 classes reachable
  12,527 (10.62%) of 117,926 fields reachable
  38,243 (55.40%) of 69,033 methods reachable
     663 classes, 1,822 fields, and 3,761 methods registered for reflection
      72 classes,    80 fields, and    60 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
[3/7] Building universe...                                                                               (1.4s @ 5.08GB)
[4/7] Parsing methods...      [*]                                                                        (1.0s @ 3.69GB)
Exception in thread "native-image pid watcher" java.lang.OutOfMemoryError: GC overhead limit exceeded
```

The steps to reproduce the error are:

1) Set `JAVA_HOME`, clone the repo and build
```
$ export JAVA_HOME=/path/to/graalvm-ce-java17-22.3.0
$ git clone git@github.com:ionutbalosin/jvm-performance-benchmarks.git

$ cd jvm-performance-benchmarks
$ ./mvnw -P jdk17_profile clean package
```

2) Run the tracing agent to build the reflection configuration file
```
$ $JAVA_HOME/bin/java --add-modules=jdk.incubator.vector -Xss64M -agentlib:native-image-agent=config-output-dir=META-INF/native-image -jar benchmarks/target/benchmarks.jar -foe true -wi 0 -w 0 -i 1 -r 0 -f 0
```

3) Compile with native image
```
$ $JAVA_HOME/bin/native-image -J-Xmx32G --add-modules=jdk.incubator.vector --initialize-at-build-time=org.openjdk.jmh.infra,org.openjdk.jmh.util.Utils,org.openjdk.jmh.runner.InfraControl,org.openjdk.jmh.runner.InfraControlL0,org.openjdk.jmh.runner.InfraControlL1,org.openjdk.jmh.runner.InfraControlL2,org.openjdk.jmh.runner.InfraControlL3,org.openjdk.jmh.runner.InfraControlL4 -H:+ReportExceptionStackTraces -jar benchmarks/target/benchmarks.jar -cp . graal17ce-image
```

<br><br> Running the tracing agent for a single benchmark seems to fix the OOM issue:  
```
$ $JAVA_HOME/bin/java --add-modules=jdk.incubator.vector -Xss64M -agentlib:native-image-agent=config-output-dir=META-INF/native-image -jar benchmarks/target/benchmarks.jar .*InfrastructureBaselineBenchmark.* -foe true -wi 0 -w 0 -i 1 -r 0 -f 0
```

Therefore, I am asking the following questions:
1) Is the tracing agent supposed to be used as shown above?
2) Given the output of step `2/7 Performing analysis` during native image compilation, are memory requirements expected to exceed 32GB?

**Comment by christianwimmer** [comment]
Thanks for the report, that was an interesting problem. The error was during method inlining, which went out of control and resulted in graphs that either no longer fit into the heap or then later on crashed compilation.

The root cause is in the vector API of the JDK, which uses the annotation `jdk.internal.vm.annotation.ForceInline` on a lot of methods, expecting that a whole call chain is inlined. But since there are virtual calls in that chain where inlining is not possible, the forced inlining of "half the call chain" gets out of control because memory loads that are expected to be constant folded are no longer constant.

I am adding a workaround to ignore the `ForceInline` annotation in such cases.

However, image build then still fails because the Vector API is not supported yet in Native Image (it is an incubator feature in the JDK) - you will get a linking error like `undefined reference to `Java_jdk_internal_vm_vector_VectorSupport_getMaxLaneCount'`.

So in summary, you need to exclude all Vector API tests, i.e., remove the `--add-modules=jdk.incubator.vector`, from the agent run and the native-image command line, and disable all tests that rely on that module.
